sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/transforms/ParDoTest.java

```diff
@@ -1523,20 +1528,66 @@ public void testMutatingInputCoderDoFnError() throws Exception {
   }
 
   @Test
-  public void testIncludesDoFnDisplayData() {
-    Bound<String, String> parDo =
-        ParDo.of(
-            new DoFn<String, String>() {
-              @Override
-              public void processElement(ProcessContext c) {}
+  public void testDoFnDisplayData() {
+    DoFn<String, String> fn = new DoFn<String, String>() {
+      @Override
+      public void processElement(ProcessContext c) {
+      }
 
-              @Override
-              public void populateDisplayData(Builder builder) {
-                builder.add("foo", "bar");
-              }
-            });
+      @Override
+      public void populateDisplayData(Builder builder) {
+        builder.add("doFnMetadata", "bar");
+      }
+    };
+
+    Bound<String, String> parDo = ParDo.of(fn);
+
+    DisplayData displayData = DisplayData.from(parDo);
+    assertThat(displayData, hasDisplayItem(allOf(
+        hasKey("fn"),
+        hasType(DisplayData.Type.JAVA_CLASS),
+        hasValue(fn.getClass().getName()))));
+
+    assertThat(displayData, includes(fn));
+  }
+
+  @Test
+  public void testDoFnWithContextDisplayData() {
+    DoFnWithContext<String, String> fn = new DoFnWithContext<String, String>() {
+      @ProcessElement
+      public void proccessElement(ProcessContext c) {}
+
+      @Override
+      public void populateDisplayData(Builder builder) {
+        builder.add("fnMetadata", "foobar");
+      }
+    };
+
+    Bound<String, String> parDo = ParDo.of(fn);
+
+    DisplayData displayData = DisplayData.from(parDo);
+    assertThat(displayData, includes(fn));
+    assertThat(displayData, hasDisplayItem("fn", fn.getClass()));
+  }
+
+  @Test
+  public void testWithOutputTagsDisplayData() {
+    DoFnWithContext<String, String> fn = new DoFnWithContext<String, String>() {
+      @ProcessElement
+      public void proccessElement(ProcessContext c) {}
+
+      @Override
+      public void populateDisplayData(Builder builder) {
+        builder.add("fnMetadata", "foobar");
+      }
+    };
+
+    ParDo.BoundMulti parDo = ParDo
+            .withOutputTags(new TupleTag(), TupleTagList.empty())
+            .of(fn);
 
     DisplayData displayData = DisplayData.from(parDo);
-    assertThat(displayData, hasDisplayItem(hasKey("foo")));
+    assertThat(displayData, includes(fn));
```

Comment:
I like the testing predicates! I assume that `includes(fn)` makes sure that all the contents of `fn` is included, and properly namespaced based on `fn`?